llvm/labs/lab2/soloninko_andrey/instrumentation.cpp

```diff
@@ -0,0 +1,74 @@
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+struct InstrumentationPass : llvm::PassInfoMixin<InstrumentationPass> {
+  llvm::PreservedAnalyses run(llvm::Function &F,
+                              llvm::FunctionAnalysisManager &) {
+    llvm::LLVMContext &context = F.getContext();
+    llvm::IRBuilder<> builder(context);
+    llvm::Module *module = F.getParent();
+    llvm::FunctionType *type =
+        llvm::FunctionType::get(llvm::Type::getVoidTy(context), false);
+    llvm::FunctionCallee func_f =
+        (*module).getOrInsertFunction("instrument_start", type);
+    llvm::FunctionCallee func_l =
+        (*module).getOrInsertFunction("instrument_end", type);
+
+    bool foundInstrument_start = false;
+    bool foundInstrument_end = false;
+
+    for (llvm::BasicBlock &BB : F) {
+      for (llvm::Instruction &I : BB) {
+        if (auto *callInst = llvm::dyn_cast<llvm::CallInst>(&I)) {
```

Comment:
In generic case it is more performant to just check that `F` is not present among users of `func_f` and `func_l` functions.